mlir/lib/Dialect/SCF/Transforms/ForallToParallel.cpp

```diff
@@ -0,0 +1,82 @@
+//===- ForallToParallel.cpp - scf.forall to scf.parallel loop conversion --===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Transforms SCF.ForallOp's into SCF.ParallelOps's.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/Passes.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/IR/PatternMatch.h"
+
+namespace mlir {
+#define GEN_PASS_DEF_SCFFORALLTOPARALLELLOOP
+#include "mlir/Dialect/SCF/Transforms/Passes.h.inc"
+} // namespace mlir
+
+using namespace mlir;
+
+LogicalResult mlir::scf::forallToParallelLoop(RewriterBase &rewriter,
```

Comment:
I think the `mapping` attribute in `forall` could (should?) be preserved.

Reply:
I had a look through the other `SCF->*` conversions to see if the `mapping` attribute was used anywhere. The `SCF->GPU` conversion specifically looks for the `mapping` attribute.

I think preserving the attribute is reasonable, though maybe the scf.parallel definition should be modified to make it an official argument.

Reply:
> I think preserving the attribute is reasonable, though maybe the scf.parallel definition should be modified to make it an official argument.

Good point. And that would remove the need for the magic naming string `gpu::getMappingAttrName()`.

Btw, could you also add a test case that shows this attribute preservation?

Reply:
I've added a test to lock down preserving the attribute. Formalizing the `mapping` attribute on the `scf.parallel` op has some fallout on the GPU dialect, so I opted not to do so. Some of the GPU transforms use the `mapping` attribute with values other than `DeviceMappingArrayAttr`, which is what `scf.forall` requires.